cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1142,82 +1145,67 @@ struct TreeWalker : public std::enable_shared_from_this<TreeWalker> {
         recursion_handler_(std::move(recursion_handler)) {}
 
  private:
-  std::mutex mutex_;
-  Future<> future_;
-  std::atomic<int32_t> num_in_flight_;
+  std::shared_ptr<TaskGroup> task_group_;
+  util::Mutex mutex_;
```

Comment:
Is there a reason not to use `std::mutex` here?

Reply:
Nope.  For some reason I thought the old implementation was using it.  I've switched it back to std::mutex.